python_modules/dagster/dagster/core/execution_plan/create.py

```diff
@@ -244,23 +253,22 @@ def _create_new_steps_for_input(state, step, subset_info):
     new_steps = []
     new_step_inputs = []
     for step_input in step.step_inputs:
-        if step_input.name in subset_info.inputs[step.key]:
-            value_thunk_step_output_handle = create_value_thunk_step(
-                state,
-                step.solid,
-                step_input.runtime_type,
-                step.key + '.input.' + step_input.name + '.value',
-                subset_info.inputs[step.key][step_input.name],
-            )
-
-            new_value_step = value_thunk_step_output_handle.step
+        if subset_info.has_injected_step(step.key, step_input.name):
+            subset_info.step_factory_fns[step.key][step_input.name](state, step, step_input)
 
-            new_steps.append(new_value_step)
+            step_output_handle = check.inst(
+                subset_info.step_factory_fns[step.key][step_input.name](state, step, step_input),
+                StepOutputHandle,
+                'Step factory function must create StepOutputHandle',
+            )
 
+            new_steps.append(step_output_handle.step)
             new_step_inputs.append(
-                StepInput(step_input.name, step_input.runtime_type, value_thunk_step_output_handle)
+                StepInput(step_input.name, step_input.runtime_type, step_output_handle)
             )
+
         else:
+            # TODO: I don't think you should able to get here
```

Comment:
if this is a bad state, should we throw?

Reply:
good catch